sklearn/decomposition/ksvd.py

```diff
@@ -0,0 +1,343 @@
+"""K-SVD dictionary learning algorithm.
+"""
+
+
+import numpy
+import scipy.sparse.linalg
+
+from ..base import BaseEstimator
+from .dict_learning import sparse_encode, SparseCodingMixin
+from ..preprocessing import normalize
+from ..utils import check_array, check_random_state
+from ..utils.extmath import norm
+from ..utils.random import choice
+
+
+# atoms with norm below are considered zero
+_ATOM_NORM_TOLERANCE = 1e-6
+
+
+def _worst_represented_example(X, dictionary, sparse_codes):
+    """Find the sample that has the most representation error."""
+
+    residuals = X - numpy.dot(sparse_codes, dictionary)
+    errors_squared = numpy.sum(residuals * residuals, axis=1)
+    worst_index = numpy.argmax(errors_squared)
+    return X[worst_index, :]
+
+
+def _decompose_svd(residual):
+    """Decompose the residual using 1-rank SVD."""
+
+    # Sparse version is used because non-sparse does not allow to
+    # set the desired number of singular values.
+    U, S, V = scipy.sparse.linalg.svds(residual, 1)
```

Comment:
If you need the vector, probably yes - it uses the power method IIRC

On Fri, Aug 28, 2015 at 1:53 PM, Arthur Mensch <notifications@github.com>
wrote:

> In sklearn/decomposition/ksvd.py
> <https://github.com/scikit-learn/scikit-learn/pull/3739#discussion_r38192315>
> :
>
> > +
> > +def _worst_represented_example(X, dictionary, sparse_codes):
> > +    """Find the sample that has the most representation error."""
> > +
> > +    residuals = X - numpy.dot(sparse_codes, dictionary)
> > +    errors_squared = numpy.sum(residuals * residuals, axis=1)
> > +    worst_index = numpy.argmax(errors_squared)
> > +    return X[worst_index, :]
> > +
> > +
> > +def _decompose_svd(residual):
> > +    """Decompose the residual using 1-rank SVD."""
> > +
> > +    # Sparse version is used because non-sparse does not allow to
> > +    # set the desired number of singular values.
> > +    U, S, V = scipy.sparse.linalg.svds(residual, 1)
>
> Is that the most efficient way to do a 1-rank decomposition of a matrix ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3739/files#r38192315>.
>